src/main/scala/chisel3/ModuleChoice.scala

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import scala.language.experimental.macros
+import scala.collection.immutable.ListMap
+
+import chisel3.FixedIOBaseModule
+import chisel3.experimental.{BaseModule, SourceInfo}
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl._
+import chisel3.internal.sourceinfo.InstChoiceTransform
+import chisel3.internal.{Builder}
+
+object ModuleChoice extends SourceInfoDoc {
+
+  /** A wrapper method for Module instantiation based on option choices.
+    * (necessary to help Chisel track internal state).
+    *
+    * @param default the Module to instantiate if no module is specified
+    * @param choices the mapping from cases to module generators
+    *
+    * @return the input module `m` with Chisel metadata properly set
+    *
+    * @throws IllegalArgumentException if the cases do not belong to the same option.
+    */
+  def apply[T <: Data](
+    default: => FixedIOBaseModule[T]
+  )(choices: => ListMap[option.Case, () => FixedIOBaseModule[T]]
```

Comment:
You're using `ListMap` in order to guarantee an ordering here. The alternative is to use a `Seq` and put the checking of uniqueness when this method is called (as opposed to putting it in the type). There are already _other_ run time checks, so this seems fine to delay similarly. I'd suggest changing this. @jackkoenig: do you have any thoughts here?

`ListMap` is just weird as it's very rarely what you want do to its terrible performance characteristics (it is not an ordered map, it is literally just a linked list that will iterate over the list to find the key). Scala 2 has no immutable ordered list, only mutable ordered lists.